cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java

```diff
@@ -90,65 +99,661 @@
-    public static void after() {
+    public static void afterAll() {
         ccServer.stop();
     }
 
     @BeforeEach
-    public void resetServer() {
+    public void beforeEach(Vertx vertx) {
         ccServer.reset();
+
+        kubernetesClient = new MockKube()
+                .withCustomResourceDefinition(Crds.kafkaRebalance(), KafkaRebalance.class, KafkaRebalanceList.class, DoneableKafkaRebalance.class)
+                .end()
+                .build();
+
+        ResourceOperatorSupplier supplier = ResourceUtils.supplierWithMocks(true);
+        PlatformFeaturesAvailability pfa = new PlatformFeaturesAvailability(true, kubernetesVersion);
+        kcrao = new KafkaRebalanceAssemblyOperator(vertx, pfa, supplier, HOST);
+
+        mockRebalanceOps = supplier.kafkaRebalanceOperator;
+        mockKafkaOps = supplier.kafkaOperator;
+    }
+
+    @AfterEach
+    public void afterEach() {
+        if (kubernetesClient != null) {
+            kubernetesClient.close();
+        }
     }
 
     @Test
-    public void testNewRebalance(Vertx vertx, VertxTestContext context) throws IOException, URISyntaxException {
+    public void testNewToProposalReadyRebalance(VertxTestContext context) throws IOException, URISyntaxException {
 
-        // Setup the rebalance user tasks endpoints with the number of pending calls before a response is received.
-        MockCruiseControl.setupCCRebalanceResponse(ccServer);
-        MockCruiseControl.setupCCUserTasksResponse(ccServer, 0);
+        // Setup the rebalance endpoint with the number of pending calls before a response is received.
+        MockCruiseControl.setupCCRebalanceResponse(ccServer, 0);
 
-        KubernetesClient client = new MockKube()
-                .withCustomResourceDefinition(Crds.kafkaRebalance(), KafkaRebalance.class, KafkaRebalanceList.class, DoneableKafkaRebalance.class)
-                .end()
+        KafkaRebalance kr =
+                createKafkaRebalance(CLUSTER_NAMESPACE, CLUSTER_NAME, RESOURCE_NAME, new KafkaRebalanceSpecBuilder().build());
+
+        Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).create(kr);
+
+        when(mockKafkaOps.getAsync(CLUSTER_NAMESPACE, CLUSTER_NAME)).thenReturn(Future.succeededFuture(kafka));
+        mockRebalanceOperator(mockRebalanceOps, CLUSTER_NAMESPACE, RESOURCE_NAME, kubernetesClient);
+
+        kcrao.reconcileRebalance(
+                new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                kr).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New directly to ProposalReady (no pending calls in the Mock server)
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.ProposalReady);
+                    context.completeNow();
```

Comment:
I am not an expert on this, so a genuine question ... what's the value? I am not using multiple steps to check but just completing the test on the last handler of each chain. I have a sequential composition not parallel async executions. If I am going to use just one flag, it seems to be quite the same as leaving as it is now.

Reply:
I think that is a more than fair question - I am by no means an expert either ;P

So semantically I like checkpoints more as `context.completeNow` signals to the reader a more abrupt end to the test (this may be a case of taste)
But to me, if i see a checkpoint it signals ahead of time that I have a single (or more) processes that have to be completed and properly state  checked before the test is done, so when I see a  flag, it is normally a softer end to test. 

I also think flags + checkpoints are easier to extend, I.e. if more needs to be added to the tests we dont have to remove the context.completeNow if it no longer makes sense.

Thirdly (coming back to semantics again) I think checkpoints and flag are more in line with the async nature of futures, a flag registers with a checkpoint and once its goal is reached the context knows the test can finish, a completeNow sort of feels like sticking a system.exit at the end of a promise which feels a little odd to me.

Don't get me wrong though, this is mostly academic and doesnt effect the tests as they are now. And I dont want to impose my coding styles on others if they dont agree

Reply:
I can see all your points. In this specific case the usage of checkpoints just collapse to be the same as completeNow with no difference. Anyway, the semantic will be the same, I can do the change easily.

Reply:
Yeah I agree - I wouldn't stop approval from you leaving them in, just a convention I have been following :)